inference-engine/src/mkldnn_plugin/nodes/mkldnn_gather_nd_node.cpp

```diff
@@ -40,180 +37,114 @@ MKLDNNGatherNDNode::MKLDNNGatherNDNode(const std::shared_ptr<ngraph::Node>& op,
-
-    std::vector<size_t> srcMultipliers(_sliceRank);
-    for (size_t i = 0; i < _sliceRank ; i++)
-        srcMultipliers[i] = _dataTypeSize * getParentEdgeAt(_dataIndex)->getMemory().GetDescWithType<BlockedMemoryDesc>()->getStrides()[i + _batchDims];
-
-    const size_t batchStep = _batchStep * _dataTypeSize;
-    const size_t dataStep = _blockSize * _dataTypeSize;
-    const size_t cycles = getChildEdgeAt(0)->getMemory().GetSize() / (dataStep * _batchNum);
-    const size_t CS = cycles * _sliceRank;
-    const size_t CB = cycles * dataStep;
-    const size_t workAmount = _batchNum * cycles;
-
-    auto threadBody = [&](const int ithr, const int nthr) {
-        size_t start(0lu), end(0lu);
-        splitter(workAmount, nthr, ithr, start, end);
-        if (start >= end)
-            return;
-        size_t bStart = start / cycles;
-        size_t cStart = start % cycles;
-        size_t workCounter = start;
-
-        const uint8_t* shiftedSrcData = srcData + bStart * batchStep;
-        const int* shiftedIndices = indices + bStart * CS + cStart * _sliceRank;
-        uint8_t* shiftedDstData = dstData + bStart * CB + cStart * dataStep;
-
-        for (size_t b = bStart; b < _batchNum; b++) {
-            for (size_t j = cStart; j < cycles; j++) {
-                size_t dataIdx = 0lu;
-                for (size_t i = 0; i < _sliceRank ; i++)
-                    dataIdx += srcMultipliers[i] * shiftedIndices[i];
-                cpu_memcpy(shiftedDstData, &(shiftedSrcData[dataIdx]), dataStep);
-                shiftedDstData += dataStep;
-                shiftedIndices += _sliceRank;
-                if (++workCounter == end) {
-                    return;
-                }
-            }
-            cStart = 0;
-            shiftedSrcData += batchStep;
-        }
-    };
+MKLDNNGatherNDNode::GatherNDExecutor::GatherNDExecutor(const GatherNDAttributes& attrs) : attrs(attrs) {
+    batchSize = std::accumulate(attrs.srcDims.begin(), attrs.srcDims.begin() + attrs.batchDims, 1lu, std::multiplies<size_t>());
+    dataLength = std::accumulate(attrs.srcDims.begin() + attrs.sliceRank + attrs.batchDims, attrs.srcDims.end(), 1lu,
+                                 std::multiplies<size_t>()) * attrs.dataSize;
+    cycles = attrs.dstSize / (dataLength * batchSize);
+
+    srcBatchStride = std::accumulate(attrs.srcDims.begin() + attrs.batchDims, attrs.srcDims.end(), 1lu,
+                                     std::multiplies<size_t>()) * attrs.dataSize;
+    idxBatchStride = cycles * attrs.sliceRank;
+    dstBatchStride = cycles * dataLength;
 
-    parallel_nt(0, threadBody);
+    srcShifts.resize(attrs.sliceRank, 0);
+    for (size_t i = 0; i < attrs.sliceRank ; i++)
+        srcShifts[i] = attrs.srcStrides[i + attrs.batchDims] * attrs.dataSize;
+}
+
+void MKLDNNGatherNDNode::GatherNDExecutor::exec(const uint8_t* srcData, const int32_t* indices, uint8_t* dstData) {
+    parallel_for2d(batchSize, cycles, [&](const size_t b, const size_t j) {
```

Comment:
You significantly changed the implementation. Is there any evidence that the new one shows the same or better performance?
I am ok to merge as is, but we have to follow up that.